### PR TITLE
[Datasets] Add support for string tensor columns in `ArrowTensorArray` and `ArrowVariableShapedTensorArray`

### PR DESCRIPTION
**Pull request by scottjlee**
Signed-off-by: Scott Lee <sjl@anyscale.com>

## Why are these changes needed?
Add support for creating `ArrowTensorArrays` and `ArrowVariableShapedTensorArray` with string typed columns. The previous PR https://github.com/ray-project/ray/pull/31817 had CI test failures which were not run at PR-review time. This PR replicates the functionality of the previous PR, and additionally addresses the test failures (which only occur for Arrow 8.0+).

## Related issue number

Based off of PR https://github.com/ray-project/ray/pull/31817, with additional implementation on top. I have marked new sections of code with GitHub comments.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 1]
```diff
@@ -598,13 +665,63 @@ def _extension_scalar_to_ndarray(
...
+                )
+                data_offset = buffer_item_width * offset
+                data_buffer = raw_values.buffers()[1]
+                return np.ndarray(
+                    shape, dtype=ext_dtype, buffer=data_buffer, offset=data_offset
+                )
+            ext_dtype = raw_values.type.to_pandas_dtype()
+            return np.ndarray(
+                shape, dtype=ext_dtype, buffer=data_buffer, offset=data_offset
+            )
```
Comment (scottjlee) [inline python/ray/air/util/tensor_extensions/arrow.py:724]: Part 1 of new code from this PR (as opposed to changes from previous PR)

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 2]
```diff
@@ -124,10 +127,60 @@ def _extension_scalar_to_ndarray(
...
```
Comment (scottjlee) [inline python/ray/air/util/tensor_extensions/arrow.py:183]: Part 1 of new code from this PR (as opposed to changes from previous PR)

File: python/ray/data/tests/test_dataset_tfrecords.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def test_read_tfrecords(ray_start_regular_shared, tmp_path):
     assert dict(df.dtypes) == {
         "int64": np.int64,
         "int64_list": object,
-        "float": np.float,
+        "float": np.float_,
```
Comment (scottjlee) [inline python/ray/data/tests/test_dataset_tfrecords.py:43]: This came up when I was testing locally with Arrow 8.0+. `np.float` is [deprecated starting numpy 1.20.0](https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated), so I have replaced it with `np.float_`.

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 4]
```diff
@@ -124,10 +127,60 @@ def _extension_scalar_to_ndarray(
...
+                    offset=byte_bucket_offset,
+                )
+                # Unpack into a byte per boolean, using LSB bit-packed ordering.
+                arr = np.unpackbits(arr, bitorder="little")
+                # Interpret buffer as boolean array.
+                return np.ndarray(shape, dtype=np.bool_, buffer=arr, offset=bool_offset)
+            # Special handling of ragged string tensors
+            if pa.types.is_fixed_size_binary(raw_values.type):
+                offset = raw_values.offset
+                buffer_item_width = raw_values.type.bit_width // 8
```
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:171]: Duplicated definitions of above
```suggestion
```

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 5]
```diff
@@ -124,10 +127,60 @@ def _extension_scalar_to_ndarray(
...
+                ext_dtype = np.dtype(
+                    f"<U{raw_values.type.byte_width // NUM_BYTES_PER_UNICODE_CHAR}"
+                )
+                data_offset = buffer_item_width * offset
+                data_buffer = raw_values.buffers()[1]
```
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:176]: Same here
```suggestion
```

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 6]
```diff
@@ -124,10 +127,60 @@ def _extension_scalar_to_ndarray(
...
+                return np.ndarray(
+                    shape, dtype=ext_dtype, buffer=data_buffer, offset=data_offset
+                )
```
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:179]: Actually, could we fallthrough to the below `np.ndarray()` constructor after overriding `ext_dtype`? Should just need to move the below `ext_dtype` to the top of this function to let the string override take place.
```suggestion
```

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 7]
```diff
@@ -598,13 +665,63 @@ def _extension_scalar_to_ndarray(
...
```
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:720]: Same here
```suggestion
            if pa.types.is_fixed_size_binary(raw_values.type):
                ext_dtype = np.dtype(
                    f"<U{raw_values.type.byte_width // NUM_BYTES_PER_UNICODE_CHAR}"
                )
```

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 8]
```diff
@@ -124,10 +127,53 @@ def _extension_scalar_to_ndarray(
...
+                return np.ndarray(shape, dtype=np.bool_, buffer=arr, offset=bool_offset)
+            ext_dtype = raw_values.type.to_pandas_dtype()
+            # Special handling of ragged string tensors
+            if pa.types.is_fixed_size_binary(raw_values.type):
+                ext_dtype = np.dtype(
+                    f"<U{raw_values.type.byte_width // NUM_BYTES_PER_UNICODE_CHAR}"
+                )
+            return np.ndarray(
+                shape, dtype=ext_dtype, buffer=data_buffer, offset=data_offset
+            )
```
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:176]: Btw, it appears that this and `ArrowVariableShapedTensorType._extension_scalar_to_ndarray()` are the same after the definitions of `shape` and `raw_values`, could we consolidate most of this logic into a single utility function that lives at the bottom of this module?

Comment (scottjlee) [reply python/ray/air/util/tensor_extensions/arrow.py]: I generalized `ArrowTensorType._extension_scalar_to_ndarray` as well (which had a few variables that were slightly different). LMK what you think

**Review by clarkzinzow — APPROVED** [review]
Awesome, great consolidation of that common ndarray building code! It looks like we could also subsume most of `ArrowTensorArray._to_numpy()` into that helper as well: https://github.com/ray-project/ray/blob/f21a9e6bf672118cc9ff38815bb241e33b915ebe/python/ray/air/util/tensor_extensions/arrow.py#L392-L447

**Comment by scottjlee** [comment]
> Awesome, great consolidation of that common ndarray building code! It looks like we could also subsume most of `ArrowTensorArray._to_numpy()` into that helper as well:
> 
> https://github.com/ray-project/ray/blob/f21a9e6bf672118cc9ff38815bb241e33b915ebe/python/ray/air/util/tensor_extensions/arrow.py#L392-L447

I was looking into this as well, but I think that method has some extra logic that's not present in the other methods, for handling the offsets for a single element: https://github.com/ray-project/ray/blob/f21a9e6bf672118cc9ff38815bb241e33b915ebe/python/ray/air/util/tensor_extensions/arrow.py#L406-L415

I could further generalize the helper method to work for this, but I think it could be overgeneralized with this new addition. Thoughts? @clarkzinzow

**Comment by clarkzinzow** [comment]
@scottjlee Ah good point, I missed that on the skim! In that case, this looks good to merge!

**Pull request merged**